Exercise 1: Training a trigram model.

I will implement the trigram model using a neural network.

Unlike the bigram model, which just took in one previous character and predict the next character, the trigram model will take in two previous characters and predict the next character.

In [1]:
import gdown
gdown.download("https://drive.google.com/uc?id=1D3VgZXIVUCxv2dZSs7cGKtbu_VcPerSw", "names.txt", quiet=False)

Downloading...
From: https://drive.google.com/uc?id=1D3VgZXIVUCxv2dZSs7cGKtbu_VcPerSw
To: /content/names.txt
100%|██████████| 228k/228k [00:00<00:00, 4.39MB/s]


'names.txt'

In [2]:
names = open('names.txt', 'r').read().splitlines()

In [3]:
names[:10]

['emma',
 'olivia',
 'ava',
 'isabella',
 'sophia',
 'charlotte',
 'mia',
 'amelia',
 'harper',
 'evelyn']

In [4]:
len(names)

32033

In [5]:
from string import ascii_lowercase

characters = list(ascii_lowercase)
characters = ['.'] + characters
characters

['.',
 'a',
 'b',
 'c',
 'd',
 'e',
 'f',
 'g',
 'h',
 'i',
 'j',
 'k',
 'l',
 'm',
 'n',
 'o',
 'p',
 'q',
 'r',
 's',
 't',
 'u',
 'v',
 'w',
 'x',
 'y',
 'z']

In [6]:
# Defining the ctoi mapping -- this maps a character to it's corresponding index.

ctoi = {c: i for i, c in enumerate(characters)}
ctoi

{'.': 0,
 'a': 1,
 'b': 2,
 'c': 3,
 'd': 4,
 'e': 5,
 'f': 6,
 'g': 7,
 'h': 8,
 'i': 9,
 'j': 10,
 'k': 11,
 'l': 12,
 'm': 13,
 'n': 14,
 'o': 15,
 'p': 16,
 'q': 17,
 'r': 18,
 's': 19,
 't': 20,
 'u': 21,
 'v': 22,
 'w': 23,
 'x': 24,
 'y': 25,
 'z': 26}

In [16]:
len(ctoi)

27

In [12]:
# Defining the itoc mapping -- this maps an index to its corresponding character.
itoc = {i: c for c, i in ctoi.items()}
itoc

{0: '.',
 1: 'a',
 2: 'b',
 3: 'c',
 4: 'd',
 5: 'e',
 6: 'f',
 7: 'g',
 8: 'h',
 9: 'i',
 10: 'j',
 11: 'k',
 12: 'l',
 13: 'm',
 14: 'n',
 15: 'o',
 16: 'p',
 17: 'q',
 18: 'r',
 19: 's',
 20: 't',
 21: 'u',
 22: 'v',
 23: 'w',
 24: 'x',
 25: 'y',
 26: 'z'}

In [17]:
len(itoc)

27

In [14]:
# Defining the cptoi mapping -- this maps a character pair to it's corresponding index.

cptoi = {(c1, c2): i for i, (c1, c2) in enumerate((c1, c2) for c1 in characters for c2 in characters)}

In [15]:
len(cptoi)

729

Ex. for the first word in the training dataset ('emma'):

'emma' -> '..emma.'


Example 1: .. -> e

Example 2: .e -> m

Example 3: em -> m

Example 4: mm -> a

Example 5: ma -> .

In [20]:
import torch

# Creating the training set of trigrams.
xs, ys = [], []

for name in names:
  cs = ['.', '.'] + list(name) + ['.']
  for c1, c2, c3 in zip(cs, cs[1:], cs[2:]):
    xs.append(cptoi[(c1, c2)])
    ys.append(ctoi[c3])

xs, ys = torch.tensor(xs), torch.tensor(ys)


In [21]:
xs.shape

torch.Size([228146])

In [22]:
ys.shape

torch.Size([228146])

In [23]:
assert len(xs) == len(ys)

n = len(xs)

Previously, with the bigram neural network, we simply encoded each input character as a 27-length OHE vector that was fed as input to the neural network.

What do we do with a trigram neural network?

Well, one way to do this is to simply encode the pair of input characters as a 729-length OHE vector that we will feed as input to the neural network.


Right now, xs is of shape 228146, and I want it to become of shape 228146 x 729.


In [55]:
# One-Hot Encoding (OHE)

import torch.nn.functional as F

def one_hot_encoding(raw):
  return F.one_hot(raw, num_classes = 729).float()

xs_encoded = one_hot_encoding(xs)
xs_encoded.shape

torch.Size([228146, 729])

Note: Another way to do this would be to take the first character in the pair, run it through a one-hot encoder, get the 27-length OHE vector result. Then, take the second character in the pair, run it through a one-hot encoder, get the 27-length OHE vector result. Then, take the two resulting 27-length OHE vectors, and concatenate them together to form a 54-length vector. I can try this later.

In [31]:
# Now, let's check our work.

# The first training example is: .. -> e

# Therefore,  at xs[0], we expect to see 0.

xs[0].item()



0

In [32]:
# At xs_encoded[0], we expect to see a 729-length vector that has a 1 at index 0.

for index, element in enumerate(xs_encoded[0]):
  print(index, element.item())

0 1.0
1 0.0
2 0.0
3 0.0
4 0.0
5 0.0
6 0.0
7 0.0
8 0.0
9 0.0
10 0.0
11 0.0
12 0.0
13 0.0
14 0.0
15 0.0
16 0.0
17 0.0
18 0.0
19 0.0
20 0.0
21 0.0
22 0.0
23 0.0
24 0.0
25 0.0
26 0.0
27 0.0
28 0.0
29 0.0
30 0.0
31 0.0
32 0.0
33 0.0
34 0.0
35 0.0
36 0.0
37 0.0
38 0.0
39 0.0
40 0.0
41 0.0
42 0.0
43 0.0
44 0.0
45 0.0
46 0.0
47 0.0
48 0.0
49 0.0
50 0.0
51 0.0
52 0.0
53 0.0
54 0.0
55 0.0
56 0.0
57 0.0
58 0.0
59 0.0
60 0.0
61 0.0
62 0.0
63 0.0
64 0.0
65 0.0
66 0.0
67 0.0
68 0.0
69 0.0
70 0.0
71 0.0
72 0.0
73 0.0
74 0.0
75 0.0
76 0.0
77 0.0
78 0.0
79 0.0
80 0.0
81 0.0
82 0.0
83 0.0
84 0.0
85 0.0
86 0.0
87 0.0
88 0.0
89 0.0
90 0.0
91 0.0
92 0.0
93 0.0
94 0.0
95 0.0
96 0.0
97 0.0
98 0.0
99 0.0
100 0.0
101 0.0
102 0.0
103 0.0
104 0.0
105 0.0
106 0.0
107 0.0
108 0.0
109 0.0
110 0.0
111 0.0
112 0.0
113 0.0
114 0.0
115 0.0
116 0.0
117 0.0
118 0.0
119 0.0
120 0.0
121 0.0
122 0.0
123 0.0
124 0.0
125 0.0
126 0.0
127 0.0
128 0.0
129 0.0
130 0.0
131 0.0
132 0.0
133 0.0
134 0.0
135 0.0
136 0.0
137 0.0
138 0.

What will the trigram neural network look like?


Let's recall what the bigram neural network looked like.

It consisted of an input layer, which was of size 27 (recall that the input to the bigram neural network was a 27-length OHE).

And then just an output layer, which was of size 27, which outputted the probability distribution over the next character.


Extending this to the trigram neural network, we can similarly start with:

An input layer, which will be of size 729 (recall that the input to the trigram neural network is a 729-length OHE).

And then just an output layer, which will be of size 27, which will output the probability distribution over the next character.



In [33]:
# Let's randomly generate the weights for every one of the 27 neurons in the layer.

g = torch.Generator().manual_seed(42)
W = torch.randn((729, 27), generator = g, requires_grad = True)

In [34]:
# Defining the softmax function.

def softmax(logits):
  # As Andrej said, these two lines are considered to be the 'softmax'.
  counts = logits.exp()
  probabilities = counts / counts.sum(1, keepdims = True)
  return probabilities


In [49]:
# Main training loop.

num_epochs = 10

for epoch in range(num_epochs):

  # Forward pass.
  logits = xs_encoded @ W
  probabilities = softmax(logits)
  loss = -probabilities[torch.arange(n), ys].log().mean()

  # Print loss.
  print(epoch, loss.item())

  # Backward pass.
  W.grad = None
  loss.backward()

  # Update.
  W.data -= 0.01 * W.grad

0 2.3044092655181885
1 2.3044092655181885
2 2.3044092655181885
3 2.3044090270996094
4 2.3044090270996094
5 2.3044090270996094
6 2.3044090270996094
7 2.3044090270996094
8 2.3044090270996094
9 2.3044090270996094


In [ ]:
"""

Let's consider again why this is correct for NLL:

loss = -probabilities[torch.arange(n), ys].log().mean()

probabilities is a n x 27 tensor-- there are n training inputs, and, for each of those inputs, it's giving us the model's predicted probability distribution over the 27 possible next characters.

probabilities[torch.arange(n), ys] is a way of indexing it that is the same as:

[probabilities[0, ys[0],
 probabilities[1, ys[1],
 probabilities[2, ys[2],
 ...
 probabilities[n - 1, ys[n - 1]]

This is just plucking out the probabilities the model is assigning to the correct characters...


...

"""

In [54]:
# Loss with the trigram model is at ~2.30 and still dropping, which is better than the loss with Karpathy's bigram model (~2.46).

In [140]:
# Sampling from the trigram model.

NUM_NAMES = 2

for i in range(NUM_NAMES):

  name = ['.', '.']

  while True:

    # Get the input.
    x = cptoi[(name[-2], name[-1])]

    # Convert to a one-hot-encoding.
    x_encoded = one_hot_encoding(torch.tensor([x]))

    # Run the forward pass.
    logits = x_encoded @ W
    probabilities = softmax(logits)

    # Sample from the predicted probability distribution.
    sampled_y = torch.multinomial(probabilities, num_samples = 1, replacement = True, generator = g).item()

    # Append the sampled character to the name.
    name.append(itoc[sampled_y])

    if sampled_y == 0:
      # Just predicted a '.', so the name is over.
      break


  print(f"Sampled name {i}: {''.join(name[2:-1])}")

Sampled name 0: jozel
Sampled name 1: adio


Exercise 2: Splitting the dataset randomly into an 80/10/10 split.

Training dataset: 80%.

Development dataset: 10%.

Testing dataset: 10%.

In [132]:
# One way to do this is by using torch.randperm(n).

permuted_indices = torch.randperm(n)

# permuted_indices is now some random permutation of the indices 0 -> n - 1.

# Now, we must split into 80/10/10.

start_of_dev = int(0.8 * n)
start_of_test = int(0.9 * n)

train_indices = permuted_indices[:start_of_dev]
dev_indices = permuted_indices[start_of_dev:start_of_test]
test_indices = permuted_indices[start_of_test:]

xs_encoded_train, ys_train = xs_encoded[train_indices], ys[train_indices]
xs_encoded_dev, ys_dev = xs_encoded[dev_indices], ys[dev_indices]
xs_encoded_test, ys_test = xs_encoded[test_indices], ys[test_indices]

In [133]:
print(xs_encoded_train.shape, ys_train.shape)
print(xs_encoded_dev.shape, ys_dev.shape)
print(xs_encoded_test.shape, ys_test.shape)

torch.Size([182516, 729]) torch.Size([182516])
torch.Size([22815, 729]) torch.Size([22815])
torch.Size([22815, 729]) torch.Size([22815])


In [ ]:
# Then proceed to train on the training set and then do the dev and test sets (trivial)...

Exercise 3: Playing with the regularization of the trigram model.

In [ ]:
# Skip (for now) (trivial)...

Exercise 4: Getting rid of one-hot encoding (OHE).

In [ ]:
# Skip (for now) (trivial)...

Exercise 5: Using cross-entropy loss instead.

What is F.cross_entropy?

Currently, in the forward pass, we are computing the softmax and then the NLL in two separate stages. This new function combines both so that we compute both of them in a single stage.




In [ ]:
# We can, thus, replace it with F.cross_entropy and we should get the same result (trivial)...